src/main/java/com/example/capstone_project/repository/impl/DepartmentRepositoryImpl.java

```diff
@@ -0,0 +1,60 @@
+package com.example.capstone_project.repository.impl;
+
+import com.example.capstone_project.entity.Department;
+
+import com.example.capstone_project.repository.CustomDepartmentRepository;
+import jakarta.persistence.EntityGraph;
+import jakarta.persistence.EntityManager;
+import jakarta.persistence.PersistenceContext;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.Sort;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+
+@Repository
+public class DepartmentRepositoryImpl implements CustomDepartmentRepository {
+    @PersistenceContext
+    private EntityManager entityManager;
+
+    @Override
+    public List<Department> getDepartmentWithPagination(String query, Pageable pageable) {
+        // HQL query
+        String hql = "select department from Department department " +
+                "where department.name like :query " +
+                "and department.isDelete = false " +
+                "order by ";
+
+        // Handling sort by and sort type
+        List<Sort.Order> sortOrderList = pageable.getSort().get().toList();
+        for (int i = 0; i < sortOrderList.size(); i++) {
+            Sort.Order order = sortOrderList.get(i);
+
+            String sortType = order.getDirection().isAscending() ? "asc" : "desc";
+            switch (order.getProperty().toLowerCase()) {
+                case "name":
+                    hql += "department.name " + sortType;
+                    break;
+                default:
+                    hql += "department.id " + sortType;
+            }
+
+            if (i != sortOrderList.size() - 1) {
+                hql += ", ";
+            } else {
+                hql += " ";
+            }
+        }
+        EntityGraph<Department> entityGraph = entityManager.createEntityGraph(Department.class);
+
+        // Thêm các thuộc tính cần nạp vào EntityGraph
+        entityGraph.addAttributeNodes("id", "name", "isDelete");
```

Comment:
Thử bỏ entityGraph xem mọi thứ có chạy bình thường không. entityGraph này dùng cho join. 

*Ví dụ khi lấy ra một list user và department của user đó, không làm gì đặc biệt thì Hibernate sẽ chạy thế này:*
- *1 câu query cho 1 list user*
- *10 câu query để lấy 10 department tương ứng cho 10 user đó rồi ghép vào với nhau*

*Với entityGraph.addAttitribute(User_.DEPARTMENT) thì Hibernate sẽ:*
- *1 câu query cho 1 list user join với department*

*Trong trường hợp này mình không join gì nên không cần entityGraph.*

Reply:
thanks ... thanks for ur sacrifice